File: assignments/assignment-1-solution/src-java/src/main/java/edu/kennesaw/crazy8s/game/GameContext.java
package edu.kennesaw.crazy8s.game;

import edu.kennesaw.crazy8s.ProgramContext;
import edu.kennesaw.crazy8s.carddeck.Deck;
import edu.kennesaw.crazy8s.carddeck.DiscardPile;
import edu.kennesaw.crazy8s.player.Players;
import java.util.Random;

/**
 * Holds shared game state and configuration values.
 */
public class GameContext {
    private final ProgramContext programContext;
    private final Deck deck;
    private final DiscardPile discardPile;
    private final Players players;
    private final boolean showAllHands;
    private int turnNumber;

    public GameContext(ProgramContext programContext, Deck deck, DiscardPile discardPile, Players players) {
        this.programContext = programContext;
        this.deck = deck;
        this.discardPile = discardPile;
        this.players = players;
        this.showAllHands = programContext.isShowAllHands();
    }

    public Random getRandomNumberGenerator() {
        return programContext.getRandomNumberGenerator();
    }

    public boolean isShowAllHands() {
        return showAllHands;
    }

    public String getGameTitle() {
        return "Crazy Eights (Simplified)";
    }

    public Deck getDeck() {
        return deck;
    }

    public DiscardPile getDiscardPile() {
        return discardPile;
    }

    public Players getPlayers() {
        return players;
    }

    public int getHandSize() {
        return programContext.getHandSize();
    }

    public int getTurnNumber() {
        return turnNumber;
    }

    public int incrementTurn() {
        turnNumber += 1;
        return turnNumber;
    }
}

File: assignments/assignment-1-solution/src-csharp/CrazyEights/Game/GameContext.cs
using CrazyEights.CardDeck;
using CrazyEights.Player;

namespace CrazyEights.Game;

/// <summary>
/// Holds shared game state and configuration values.
/// </summary>
public class GameContext
{
    private readonly ProgramContext _programContext;

    public GameContext(ProgramContext programContext, Deck deck, DiscardPile discardPile, Players players)
    {
        _programContext = programContext;
        Deck = deck;
        DiscardPile = discardPile;
        Players = players;
        ShowAllHands = programContext.ShowAllHands;
    }

    public Random RandomNumberGenerator => _programContext.RandomNumberGenerator;

    public bool ShowAllHands { get; }

    public string GameTitle { get; } = "Crazy Eights (Simplified)";

    public Deck Deck { get; }

    public DiscardPile DiscardPile { get; }

    public Players Players { get; }

    public int HandSize => _programContext.HandSize;

    public int TurnNumber { get; private set; }

    public int IncrementTurn()
    {
        return TurnNumber += 1;
    }
}
